Initial documentation for modular telemetry (#175)
kyma-project/telemetry-manager: docs/user/README.md

# Kyma Telemetry Module

Fundamentally, ["Observability"](https://opentelemetry.io/docs/concepts/observability-primer/) is a measure of how well the application's external outputs can reflect the internal states of single components. The insights that an application and the surrounding infrastructure expose are displayed in the form of metrics, traces, and logs - collectively, that's called "telemetry" or ["signals"](https://opentelemetry.io/docs/concepts/signals/). These can be exposed by employing modern instrumentation.

In order to implement Day-2 operations for a distributed application running in a container runtime, the single components of an application must expose these signals by employing modern instrumentation. Furthermore, the signals must be collected and enriched with the infrastructural metadata in order to ship them to a target system.

![Stages of Observability](./assets/general-stages.drawio.svg)

> There are plenty of observability backends available either as a service or as a self-manageable solution focusing on different aspects and scenarios. One approach will never fit all requirements, and the need to integrate with a specific backend will always be present. That's why Kyma supports you with instrumenting and shipping your telemetry data easily in a vendor-neutral way to enable observability for your application with little effort by integrating it into existing backends. These aspects must happen alongside your application, and managed tooling with guidance can benefit your initial investment and maintenance effort. Also, Kyma does not focus on providing a managed in-cluster backend solution, because a solution for an enterprise-grade setup demands a central out-cluster solution.

Reviewer: Trying to rephrase it with less justification, more benefits:

```suggestion
Instead of providing a one-size-for-all backend solution, Kyma supports you with instrumenting and shipping your telemetry data in a vendor-neutral way. This way, you can conveniently enable observability for your application by integrating it into your existing or desired backends. Pick your favourite among many observability backends, available either as a service or as a self-manageable solution, that focus on different aspects and scenarios. If you configure your telemetry tooling alongside your application, managed tooling with guidance can benefit your initial investment and maintenance effort. An enterprise-grade setup demands a central solution outside the cluster, so we recommend in-cluster solutions only for testing purposes.
```

The Telemetry module focuses exactly on these aspects (instrumentation/collection/shipment) happening in the runtime and explicitly defocuses on backends. In addition, you can check the [tutorials](05-tutorials.md) on how to install lightweight in-cluster backends for demo or development purposes.

## Features

The Telemetry module enables your application with the telemetry support by providing the following:

- guidance for the instrumentation - based on [Open Telemetry](https://opentelemetry.io/), community samples are provided on how to instrument your code using the [Open Telemetry SDKs](https://opentelemetry.io/docs/instrumentation/) in nearly any programming language

> - tooling for collection, filtering, and shipment - based on the [Open Telemetry Collector](https://opentelemetry.io/docs/collector/), you can configure basic pipelines to filter and ship telemetry data

Reviewer: Same ambiguity as in line 18. I'm guessing the punctuation here, maybe the other option is correct? 

```suggestion
- Tooling for collection, filtering, and shipment: Based on the [Open Telemetry Collector](https://opentelemetry.io/docs/collector/), you can configure basic pipelines to filter and ship telemetry data.
```

> - integration in a vendor-neutral way to a vendor-specific observability system - based on the [OpenTelemetry protocol (OTLP)](https://opentelemetry.io/docs/reference/specification/protocol/), backend systems can be integrated

Reviewer: Same ambiguity as above.

```suggestion
- Integration in a vendor-neutral way to a vendor-specific observability system: Based on the [OpenTelemetry protocol (OTLP)](https://opentelemetry.io/docs/reference/specification/protocol/), you can integrate backend systems.
```

- opt-out from features for advanced scenarios - at any time you can opt out per data type and execute the telemetry data collection/shipment with custom tooling
- SAP BTP as first-class integration - integration into BTP Observability services will be prioritized

> - Enterprise-grade qualities - the setup is battle-tested and will satisfy typical development standards

Reviewer: ```suggestion
- Enterprise-grade qualities: The setup is battle-tested and satisfies typical development standards.
```

This sounds like marketing-blurb phrases. What exactly does this mean?? 
Shouldn't it go without saying that we satisfy typical dev standards? That's kind of a low bar?

## Scope

> There will be a focus on the signals of application logs, distributed traces, and metrics only. Other kinds of signals are not considered. Also, logs like audit logs or operational logs are specifically descoped.

Reviewer: ```suggestion
The Telemetry module focuses only on the signals of application logs, distributed traces, and metrics. Other kinds of signals are not considered. Also, logs like audit logs or operational logs are not in scope.
```

> Supported integration scenarios are neutral to the vendor of the target system; a vendor-specific way is not planned at the moment.

Reviewer: ```suggestion
Supported integration scenarios are neutral to the vendor of the target system.
```

Roadmap information is not part of user docs.

> Initially,

Reviewer: ```suggestion
Currently, logs are based on the Fluent Bit protocol. If you're curious about the progress to switch to the vendor-neutral OTLP protocol, follow this [epic](https://github.com/kyma-project/kyma/issues/16307).
```

> - logs will not be based on the vendor-neutral OTLP protocol; follow this [epic](https://github.com/kyma-project/kyma/issues/16307) to understand the current progress on that

Reviewer: ```suggestion
```

## Components

![Components](./assets/general-components.drawio.svg)

> ### Telemetry Operator

Reviewer: Here and in the following lines: Operator or manager?

> ### Telemetry Operator

Author: ```suggestion
### Telemetry Manager
```

> The module ships the Telemetry Operator as its heart component. The operator implements the Kubernetes controller pattern and manages the whole lifecycle of all other components covered in the module. The operator watches for LogPipeline, TracePipeline, and MetricPipeline resources created by the user. With these, the user describes in a declarative way what data of a signal type to collect and where to ship it.

Reviewer: ```suggestion
The module ships the Telemetry Operator as its core component. The operator implements the Kubernetes controller pattern and manages the whole lifecycle of all other components covered in the module. The operator watches for LogPipeline, TracePipeline, and MetricPipeline resources created by the user. With these, the user describes in a declarative way what data of a signal type to collect and where to ship it.
```

This paragraph appears in another document as well, right? Make sure all the changes applied to one document are also apply to the other doc.

> If the operator detects a configuration, it will on demand roll out the relevant collector components.

Reviewer: ```suggestion
If the operator detects a configuration, it rolls out the relevant collector components on demand.
```

> You can find more details about the operator on the [operator](./01-manager.md) page.

Author: ```suggestion
You can find more details about the manager on the [manager](./01-manager.md) page.
```

> ### Log Agent

Reviewer: Here and in the following lines: agent or collector?

> The log agent is based on a [Fluent Bit](https://fluentbit.io/) installation running as a [DaemonSet](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/). It reads all containers' logs in the runtime and ships them according to a LogPipeline configuration. More details can be found in the detailed section about [Logs](./02-logs.md).

Reviewer: ```suggestion
The log agent is based on a [Fluent Bit](https://fluentbit.io/) installation running as a [DaemonSet](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/). It reads all containers' logs in the runtime and ships them according to a LogPipeline configuration. For details, see [Logs](./02-logs.md).
```

### Trace Gateway

The trace gateway is based on a [OTel Collector](https://opentelemetry.io/docs/collector/) [Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/). It provides an [OTLP-based](https://opentelemetry.io/docs/reference/specification/protocol/) endpoint where applications can push the trace signals. According to a TracePipeline configuration, the collector will process and ship the trace data to a target system. More details can be found in the detailed section about [Traces](./03-traces.md).

### Metric Gateway/Agent

> The metric gateway and agent is based on a [OTel Collector](https://opentelemetry.io/docs/collector/) [Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) and a [DaemonSet](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/). It provides an [OTLP-based](https://opentelemetry.io/docs/reference/specification/protocol/) endpoint where applications can push the metric signals. Also provides scraping of annotated prometheus based workloads. According to a MetricPipeline configuration, the collector will process and ship the metric data to a target system. More details can be found in the detailed section about [Traces](./04-metrics.md).

Reviewer: ```suggestion
The metric gateway and agent is based on a [OTel Collector](https://opentelemetry.io/docs/collector/) [Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) and a [DaemonSet](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/). It provides an [OTLP-based](https://opentelemetry.io/docs/reference/specification/protocol/) endpoint to which applications can push the metric signals. It also scrapes annotated Prometheus-based workloads. According to a MetricPipeline configuration, the collector processes and ships the metric data to a target system. For details, see [Traces](./04-metrics.md).
```
